CMakeLists.txt

```diff
@@ -546,6 +514,59 @@ if(CURL_USE_NSS)
   endif()
 endif()
 
+# Keep ZLIB detection after TLS detection,
+# and before calling CheckQuicSupportInOpenSSL.
+
+set(HAVE_LIBZ OFF)
+set(USE_ZLIB OFF)
+optional_dependency(ZLIB)
+if(ZLIB_FOUND)
+  set(HAVE_LIBZ ON)
+  set(USE_ZLIB ON)
+
+  # Depend on ZLIB via imported targets if supported by the running
+  # version of CMake.  This allows our dependents to get our dependencies
+  # transitively.
+  if(NOT CMAKE_VERSION VERSION_LESS 3.4)
+    list(APPEND CURL_LIBS ZLIB::ZLIB)
+  else()
+    list(APPEND CURL_LIBS ${ZLIB_LIBRARIES})
+    include_directories(${ZLIB_INCLUDE_DIRS})
+  endif()
+  list(APPEND CMAKE_REQUIRED_INCLUDES ${ZLIB_INCLUDE_DIRS})
+endif()
+
+option(CURL_BROTLI "Set to ON to enable building curl with brotli support." OFF)
+set(HAVE_BROTLI OFF)
+if(CURL_BROTLI)
+  find_package(Brotli QUIET)
+  if(BROTLI_FOUND)
+    set(HAVE_BROTLI ON)
+    list(PREPEND CURL_LIBS ${BROTLI_LIBRARIES})
+    list(APPEND CURL_LIBS ${BROTLI_LIBRARIES})
```

Comment:
I doubt it ever made sense to PREPEND in addition to APPEND.

Reply:
This has been a while ago, and can't recall the exact test results for this specific case. But, it's unlikely I'd add such thing unless required to fix the picky-linker problem. (A clean solution would be `-Wl,--start-group` / `-Wl,--end-group`, but could not figure out how to use with CMake correctly.)

If a better solution to this exists, I'd be very glad to learn about it. It's been a two-decade problem with this linker and static linking.

Reply:
The "picky-linker problem" is unknown to me, but I know that the order of static libs matters outside MSVC.
And I know that CMake may mess with the order unless it is properly defined by target dependencies. Subject to CMake version.
(And I know that meson still sorts link libs it gets from CMake...)

FTR I'm about  to patch out the PREPEND line in the vcpkg port. Vcpkg CI builds with static library linkage for windows, linux, osx, and android. And of course I'm just tracing a linking issue. I want to get rid of distractors.

Reply:
"Picky-linker" is my term, it applies to _binutils_ `ld` specifically. It may apply to others, but I haven't seen it anywhere else. Also llvm/clang's `lld` doesn't have this problem (neither does MSVC). My understanding is that the linker does no effort to resolve symbols in a second pass, so if a symbol is referenced earlier than defined, linking breaks. It must have a strong reason doing this for decades and without a fix.

Feel free to patch it out locally, though I'm not aware of any issues it may cause. You may also move it inside an `if(GCC)`, if it does.

Reply:
> My understanding is that the linker does no effort to resolve symbols in a second pass, so if a symbol is referenced earlier than defined, linking breaks.

It is the other way round. First the dependent lib (`-lbrotlidec`), then the independent lib (`-lbrotlicommon`). 
If you get it wrong, then you need to repeat it.

And that's why extra *prepending* is still in the "never made sense" box for me.
(Hint: `-lz` and `-lm` are typically found at the end of the link libs.)

And that's why "sort" and "remove duplicates" do not work on link libs.

Reply:
Would this help with the root cause for brotli?:
```diff
diff --git a/CMake/FindBrotli.cmake b/CMake/FindBrotli.cmake
index 11ab7f825..dcab8f7cd 100644
--- a/CMake/FindBrotli.cmake
+++ b/CMake/FindBrotli.cmake
@@ -40,4 +40,4 @@ find_package_handle_standard_args(Brotli
 )
 
 set(BROTLI_INCLUDE_DIRS ${BROTLI_INCLUDE_DIR})
-set(BROTLI_LIBRARIES ${BROTLICOMMON_LIBRARY} ${BROTLIDEC_LIBRARY})
+set(BROTLI_LIBRARIES ${BROTLIDEC_LIBRARY} ${BROTLICOMMON_LIBRARY})
```

I'm trying to replicate it, but without knowing the exact trigger configuraiton, this might take some time. If it triggers in my current test setup at all.